src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx

```diff
@@ -1519,4 +1519,13 @@
     <value>and all other platforms</value>
     <comment>This call site is reachable on: 'windows' 10.0.2000 and later, and all other platforms</comment>
   </data>
+  <data name="UseAssignableTypeForForeachVariableDescription" xml:space="preserve">
+    <value>There is no implicit conversion between collection element and iteratoi variable, this will result in runtime exception.</value>
+  </data>
+  <data name="UseAssignableTypeForForeachVariableMessage" xml:space="preserve">
+    <value>Cannot convert type '{0}' to '{1}'</value>
```

Comment:
Looking at the tests, it seems that we can report on cases where it's possible the runtime would work so we should probably be less "strict" here.

```csharp
using System.Collections.Generic;
using System;

public class A {}
public class B : A {}
public class Program
{
    public static void Main(string[] args)
    {
        var list = new List<A> { new B(), new B() };
        //Your code goes here
        
        foreach (B item in list) // We report here but in practice this code doesn't fail
            Console.WriteLine(item);
    }
}
```

Reply:
That's correct. If explicit conversion exists, but implicit one doesn't analyzer will also be reported.
We cannot guarantee that explicit conversion will succeed though. 

If explicit conversion is untended it's better for user to perform it manually.